File: app/src/main/java/com/lasercyber/lws/ui/network/http/local/CameraAiHttpPublisher.java
package com.lasercyber.lws.ui.network.http.local;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.lasercyber.lws.ai.sampling.AiFrameSamplingInterval;
import com.lasercyber.lws.ai.model.AiStainDetectResult;
import com.lasercyber.lws.ai.stain.AiStainDetectResultMapper;
import com.lasercyber.lws.ai.model.OpencvStainDetectResult;
import com.lasercyber.lws.ai.stain.OpencvStainDetectResultMapper;
import com.lasercyber.lws.ai.model.StainDetectSource;
import com.lasercyber.lws.ai.stream.StreamDetectEvent;
import com.lasercyber.lws.ai.stream.StreamDetectResultBus;
import com.lasercyber.lws.ui.common.config.CameraConfig;
import com.lasercyber.lws.ui.common.constant.LogTAGConstant;

import java.io.InputStream;
import java.util.UUID;

/**
 * Reference-counted SSE inference stream for {@code GET /v1/camera/ai}.
 */
public final class CameraAiHttpPublisher implements StreamDetectResultBus.DetectResultListener {

    private static final String TAG = LogTAGConstant.CAMERA_AI_HTTP;
    private static final CameraAiHttpPublisher INSTANCE = new CameraAiHttpPublisher();
    private final AiInferenceSseHub sseHub = AiInferenceSseHub.forLiveCamera(TAG);
    private final Object sessionLock = new Object();
    @Nullable
    private String activeSessionSource;
    private volatile boolean nativeBusSubscribed;
    private volatile int lastNativeImageWidth;
    private volatile int lastNativeImageHeight;

    private CameraAiHttpPublisher() {
        ensureNativeBusSubscription();
    }

    @NonNull
    public static CameraAiHttpPublisher getInstance() {
        INSTANCE.ensureNativeBusSubscription();
        return INSTANCE;
    }

    private void ensureNativeBusSubscription() {
        if (!CameraConfig.isNativeStreamDetectPipelineEnabled() || nativeBusSubscribed) {
            return;
        }
        StreamDetectResultBus.getInstance().addListener(this);
        nativeBusSubscribed = true;
    }

    public void publish(@NonNull AiStainDetectResult result) {
        sseHub.publishLiveCameraInference(result);
    }

  public void onInferenceSessionStart(@NonNull String source,
                                        long samplingIntervalMs,
                                        int imageWidth,
                                        int imageHeight) {
        synchronized (sessionLock) {
            if ("ai_vision_live".equals(source)
                    && StainDetectSource.LIVE.equals(activeSessionSource)) {
                return;
            }
            if (source.equals(activeSessionSource) && sseHub.hasActiveSession()) {
                return;
            }
            if (StainDetectSource.LIVE.equals(source) && activeSessionSource != null) {
                String oldId = sseHub.getActiveSessionId();
                if (oldId != null) {
                    sseHub.notifySessionStopped(
                            oldId, "preview_stopped", android.os.SystemClock.elapsedRealtime());
                }
            }
            activeSessionSource = source;
            sseHub.notifySessionStarted(new AiInferenceSseJson.SessionStart(
                    UUID.randomUUID().toString(),
                    source,
                    samplingIntervalMs,
                    imageWidth > 0 ? imageWidth : null,
                    imageHeight > 0 ? imageHeight : null));
        }
    }

    public void onInferenceSessionStop(@NonNull String reason) {
        synchronized (sessionLock) {
            if (!sseHub.hasActiveSession()) {
                activeSessionSource = null;
                return;
            }
            String sessionId = sseHub.getActiveSessionId();
            if (sessionId == null) {
                activeSessionSource = null;
                return;
            }
            activeSessionSource = null;
            sseHub.notifySessionStopped(sessionId, reason, android.os.SystemClock.elapsedRealtime());
        }
    }

    @NonNull
    public SseSubscriber acquire() {
        AiInferenceSseHub.SseSubscriber sub = sseHub.acquireSubscriber();
        CameraAiHttpActiveSignal.setCameraAiHttpSubscriberCount(sseHub.getSubscriberCount());
        return new SseSubscriber(sub);
    }

    void onSubscriberReleased() {
        CameraAiHttpActiveSignal.setCameraAiHttpSubscriberCount(sseHub.getSubscriberCount());
    }

    @Override
    public void onDetectResult(@NonNull StreamDetectEvent.DetectResult event) {
        if (!CameraConfig.isNativeStreamDetectPipelineEnabled()) {
            return;
        }
        if (!"lens_det".equals(event.module)) {
            return;
        }
        lastNativeImageWidth = event.imageWidth;
        lastNativeImageHeight = event.imageHeight;
        OpencvStainDetectResult opencvResult = OpencvStainDetectResultMapper.fromNativeSummary(
                event.summaryJson,
                event.imageWidth,
                event.imageHeight,
                event.timestampMs,
                StainDetectSource.LIVE);
        AiStainDetectResult sseResult = AiStainDetectResultMapper.processVideoStainDetectRow(
                opencvResult,
                event.imageWidth,
                event.imageHeight,
                event.timestampMs,
                StainDetectSource.LIVE);
        publish(sseResult);
    }

    @Override
    public void onSessionStart(@NonNull StreamDetectEvent.SessionStart event) {
        if (!CameraConfig.isNativeStreamDetectPipelineEnabled()) {
            return;
        }
        String source = event.source.isEmpty() ? StainDetectSource.LIVE : event.source;
        long intervalMs = event.samplingIntervalMs > 0
                ? event.samplingIntervalMs
                : AiFrameSamplingInterval.LIVE_WELD.getIntervalMs();
        onInferenceSessionStart(
                source,
                intervalMs,
                lastNativeImageWidth,
                lastNativeImageHeight);
        CameraAiHttpActiveSignal.setLivePr1InferenceStreaming(true);
    }

    @Override
    public void onSessionStop(@NonNull StreamDetectEvent.SessionStop event) {
        if (!CameraConfig.isNativeStreamDetectPipelineEnabled()) {
            return;
        }
        onInferenceSessionStop(mapNativeSessionStopReason(event.reason));
        CameraAiHttpActiveSignal.setLivePr1InferenceStreaming(false);
    }

    @Override
    public void onPipelineState(@NonNull StreamDetectEvent.PipelineState event) {
        if (!CameraConfig.isNativeStreamDetectPipelineEnabled()) {
            return;
        }
        if ("error".equals(event.state)) {
            sseHub.publishError(-1, event.detail.isEmpty() ? "stream_detect_error" : event.detail);
            CameraAiHttpActiveSignal.setLivePr1InferenceStreaming(false);
        }
    }

    @NonNull
    private static String mapNativeSessionStopReason(@NonNull String reason) {
        if ("release".equals(reason) || "laser_enable_off".equals(reason)) {
            return "laser_off";
        }
        return reason.isEmpty() ? "stream_error" : reason;
    }

    @VisibleForTesting
    static void resetForTest() {
        CameraAiHttpPublisher p = INSTANCE;
        if (p.nativeBusSubscribed) {
            StreamDetectResultBus.getInstance().removeListener(p);
            p.nativeBusSubscribed = false;
        }
        synchronized (p.sessionLock) {
            p.activeSessionSource = null;
        }
        p.lastNativeImageWidth = 0;
        p.lastNativeImageHeight = 0;
        p.sseHub.resetForTest();
        CameraAiHttpActiveSignal.setCameraAiHttpSubscriberCount(0);
        CameraAiHttpActiveSignal.setLivePr1InferenceStreaming(false);
        CameraAiHttpActiveSignal.setAiVisionLiveInferActive(false);
    }

    public static final class SseSubscriber {
        private final AiInferenceSseHub.SseSubscriber delegate;

        SseSubscriber(@NonNull AiInferenceSseHub.SseSubscriber delegate) {
            this.delegate = delegate;
        }

        @NonNull
        AiInferenceSseHub.SseSubscriber hubSubscriber() {
            return delegate;
        }

        void close() {
            delegate.closeFromClient();
            INSTANCE.onSubscriberReleased();
        }

        @NonNull
        public InputStream getInputStream() {
            return new SseInputStream(INSTANCE, delegate);
        }
    }

    private static final class SseInputStream extends InputStream {
        private final CameraAiHttpPublisher publisher;
        private final InputStream inner;

        SseInputStream(@NonNull CameraAiHttpPublisher publisher,
                       @NonNull AiInferenceSseHub.SseSubscriber delegate) {
            this.publisher = publisher;
            this.inner = delegate.getInputStream();
        }

        @Override
        public void close() throws java.io.IOException {
            inner.close();
            publisher.onSubscriberReleased();
        }

        @Override
        public int read() throws java.io.IOException {
            return inner.read();
        }

        @Override
        public int read(@NonNull byte[] buffer, int offset, int len) throws java.io.IOException {
            return inner.read(buffer, offset, len);
        }
    }
}

File: app/lws_hmi/lib/features/ai/application/camera_ai_http_publisher.dart
import 'dart:io';
import 'dart:math';

import 'package:lws_hmi/features/ai/application/ai_inference_sse_hub.dart';
import 'package:lws_hmi/features/ai/application/ai_inference_sse_json.dart';
import 'package:lws_hmi/features/ai/application/opencv_stain_detect_mapper.dart';

/// Reference-counted SSE inference stream for `GET /v1/camera/ai`.
final class CameraAiHttpPublisher {
  CameraAiHttpPublisher({
    AiInferenceSseHub? hub,
    OpencvStainDetectMapper mapper = const OpencvStainDetectMapper(),
    List<Directory> Function()? searchRoots,
  })  : _hub = hub ?? AiInferenceSseHub(),
        _mapper = mapper,
        _searchRoots = searchRoots;

  final AiInferenceSseHub _hub;
  final OpencvStainDetectMapper _mapper;
  final List<Directory> Function()? _searchRoots;

  String? _activeSessionSource;
  int _lastImageWidth = 0;
  int _lastImageHeight = 0;

  AiInferenceSseHub get hub => _hub;

  int get subscriberCount => _hub.subscriberCount;

  AiInferenceSseSubscriber acquire() {
    return _hub.acquire();
  }

  void ingestDaemonEvent(Map<String, dynamic> evt) {
    final type = evt['type']?.toString() ?? '';
    switch (type) {
      case 'detect_result':
        _onDetectResult(evt);
        break;
      case 'session_start':
        _onSessionStart(evt);
        break;
      case 'session_stop':
        _onSessionStop(evt);
        break;
      case 'pipeline_state':
        _onPipelineState(evt);
        break;
    }
  }

  void onInferenceSessionStart({
    required String source,
    required int samplingIntervalMs,
    int imageWidth = 0,
    int imageHeight = 0,
  }) {
    if (source == 'ai_vision_live' &&
        _activeSessionSource == OpencvStainDetectMapper.liveSource) {
      return;
    }
    if (source == _activeSessionSource && _hub.hasActiveSession) {
      return;
    }
    if (source == OpencvStainDetectMapper.liveSource &&
        _activeSessionSource != null) {
      final oldId = _hub.activeSessionId;
      if (oldId != null) {
        _hub.notifySessionStopped(sessionId: oldId, reason: 'preview_stopped');
      }
    }
    _activeSessionSource = source;
    final w = imageWidth > 0 ? imageWidth : _lastImageWidth;
    final h = imageHeight > 0 ? imageHeight : _lastImageHeight;
    _hub.notifySessionStarted(
      AiInferenceSessionStart(
        sessionId: _newSessionId(),
        source: source,
        samplingIntervalMs: samplingIntervalMs > 0 ? samplingIntervalMs : 500,
        imageWidth: w > 0 ? w : null,
        imageHeight: h > 0 ? h : null,
      ),
    );
  }

  void onInferenceSessionStop(String reason) {
    if (!_hub.hasActiveSession) {
      _activeSessionSource = null;
      return;
    }
    final sessionId = _hub.activeSessionId;
    _activeSessionSource = null;
    if (sessionId == null) {
      return;
    }
    _hub.notifySessionStopped(sessionId: sessionId, reason: reason);
  }

  void resetForTest() {
    _activeSessionSource = null;
    _lastImageWidth = 0;
    _lastImageHeight = 0;
    _hub.resetForTest();
  }

  void _onDetectResult(Map<String, dynamic> evt) {
    if (evt['module']?.toString() != 'lens_det') {
      return;
    }
    _lastImageWidth = _asInt(evt['imageWidth']) ?? _lastImageWidth;
    _lastImageHeight = _asInt(evt['imageHeight']) ?? _lastImageHeight;
    final source = _activeSessionSource ?? OpencvStainDetectMapper.liveSource;
    final roots = _searchRoots?.call() ?? const <Directory>[];
    final sample = _mapper.fromDetectResult(
      event: evt,
      source: source,
      searchRoots: roots,
    );
    _hub.publishRunning(sample);
  }

  void _onSessionStart(Map<String, dynamic> evt) {
    final source = (evt['source']?.toString().isNotEmpty ?? false)
        ? evt['source']!.toString()
        : OpencvStainDetectMapper.liveSource;
    final interval = _asInt(evt['samplingIntervalMs']) ?? 500;
    onInferenceSessionStart(
      source: source,
      samplingIntervalMs: interval,
      imageWidth: _lastImageWidth,
      imageHeight: _lastImageHeight,
    );
  }

  void _onSessionStop(Map<String, dynamic> evt) {
    final reason = mapNativeSessionStopReason(evt['reason']?.toString() ?? '');
    onInferenceSessionStop(reason);
  }

  void _onPipelineState(Map<String, dynamic> evt) {
    if (evt['state']?.toString() != 'error') {
      return;
    }
    final detail = evt['detail']?.toString() ?? '';
    _hub.publishError(
      code: -1,
      message: detail.isEmpty ? 'stream_detect_error' : detail,
    );
    _activeSessionSource = null;
  }

  static String mapNativeSessionStopReason(String reason) {
    if (reason == 'release' || reason == 'laser_enable_off') {
      return 'laser_off';
    }
    return reason.isEmpty ? 'stream_error' : reason;
  }

  static String _newSessionId() {
    final r = Random.secure();
    final ms = DateTime.now().millisecondsSinceEpoch;
    return '$ms-${r.nextInt(1 << 32).toRadixString(16)}';
  }

  static int? _asInt(Object? v) {
    if (v is int) return v;
    if (v is num) return v.toInt();
    return int.tryParse(v?.toString() ?? '');
  }
}
